pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -133,84 +129,172 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
-
-			rootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
-			if err = ioutil.WriteFile(internalSelfSignedRootPath, rootCert, 0o600); err != nil {
-				return err
-			}
-
+			caBundle = s.CA.GetCAKeyCertBundle().GetRootCertPem()
 			s.addStartFunc(func(stop <-chan struct{}) error {
 				go func() {
-					for {
-						select {
-						case <-stop:
-							return
-						case <-time.After(controller.NamespaceResyncPeriod):
-							newRootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
-							if !bytes.Equal(rootCert, newRootCert) {
-								rootCert = newRootCert
-								if err = ioutil.WriteFile(internalSelfSignedRootPath, rootCert, 0o600); err != nil {
-									log.Errorf("Failed to update local copy of self-signed root: %v", err)
-								} else {
-									log.Info("Updated local copy of self-signed root")
-								}
-							}
-						}
-					}
+					// regenerate istiod key cert when root cert changes.
+					s.watchRootCertAndGenKeyCert(names, stop)
 				}()
 				return nil
 			})
-			s.caBundlePath = internalSelfSignedRootPath
 		} else {
 			log.Infof("Use plugged-in cert at %v", signingKeyFile)
-			s.caBundlePath = path.Join(LocalCertDir.Get(), "root-cert.pem")
+			caBundle, err = ioutil.ReadFile(path.Join(LocalCertDir.Get(), "root-cert.pem"))
+			if err != nil {
+				return fmt.Errorf("failed reading %s: %v", path.Join(LocalCertDir.Get(), "root-cert.pem"), err)
+			}
 		}
-
 	} else {
 		log.Infof("User specified cert provider: %v", features.PilotCertProvider.Get())
 		return nil
 	}
-	if err != nil {
-		return err
+	s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
+	return nil
+}
+
+// TODO(hzxuzonghu): support async notification instead of polling the CA root cert.
+func (s *Server) watchRootCertAndGenKeyCert(names []string, stop <-chan struct{}) {
+	caBundle := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+	for {
+		select {
+		case <-stop:
+			return
+		case <-time.After(controller.NamespaceResyncPeriod):
+			newRootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+			if !bytes.Equal(caBundle, newRootCert) {
+				caBundle = newRootCert
+				certChain, keyPEM, err := s.CA.GenKeyCert(names, SelfSignedCACertTTL.Get(), false)
+				if err != nil {
+					log.Errorf("failed generating istiod key cert %v", err)
+				} else {
+					s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
+				}
+			}
+		}
 	}
+}
 
-	// Save the certificates to ./var/run/secrets/istio-dns - this is needed since most of the code we currently
-	// use to start grpc and webhooks is based on files. This is a memory-mounted dir.
-	if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-		return err
+// initCertificateWatches sets up watches for the dns certs.
+func (s *Server) initCertificateWatches(tlsOptions TLSOptions) error {
+	hasPluginCert := hasCustomTLSCerts(tlsOptions)
+	if !hasPluginCert && !s.EnableCA() && features.PilotCertProvider.Get() == IstiodCAProvider {
```

Comment:
this condition is hard to read. Anyway of simplifying it?

Reply:
will have a try to do